source/libNsdl/src/sn_grs.c

```diff
@@ -344,27 +362,147 @@ extern int8_t sn_grs_create_resource(struct grs_s *handle, sn_nsdl_resource_info
     return SN_GRS_LIST_ADDING_FAILURE;
 }
 
-int8_t sn_grs_put_resource(struct grs_s *handle, sn_nsdl_resource_info_s *res)
+/**
+ * \fn  static int8_t sn_grs_add_resource_to_list(sn_nsdl_dynamic_resource_parameters_s *resource_ptr)
+ *
+ * \brief Adds given resource to resource list
+ *
+ *  \param  *resource_ptr           Pointer to the path string to be search
+ *
+ *  \return 0 = SN_NSDL_SUCCESS, -1 = SN_NSDL_FAILURE
+ *
+*/
+static int8_t sn_grs_add_resource_to_list(struct grs_s *handle, sn_nsdl_dynamic_resource_parameters_s *resource_ptr)
+{
+    /* Local variables */
+    assert(resource_ptr->static_resource_parameters != NULL);
+
+    uint8_t *path_start_ptr = NULL;
+    uint16_t path_len = 0;
+    sn_nsdl_dynamic_resource_parameters_s *resource_copy_ptr = NULL;
+
+    /* Allocate memory for the resource info copy */
+    if (!resource_ptr->static_resource_parameters->pathlen) { //Dead code
+        return SN_NSDL_FAILURE;
+    }
+
+    resource_copy_ptr = handle->sn_grs_alloc(sizeof(sn_nsdl_dynamic_resource_parameters_s));
+    if (resource_copy_ptr == NULL) {
+        return SN_NSDL_FAILURE;
+    }
+
+    /* Set everything to zero  */
+    memset(resource_copy_ptr, 0, sizeof(sn_nsdl_dynamic_resource_parameters_s));
+    resource_copy_ptr->sn_grs_dyn_res_callback = resource_ptr->sn_grs_dyn_res_callback;
+    resource_copy_ptr->publish_uri = resource_ptr->publish_uri;
+
+    /* If resource parameters exists, copy them */
+    if (resource_ptr->static_resource_parameters) {
+        resource_copy_ptr->static_resource_parameters = handle->sn_grs_alloc(sizeof(sn_nsdl_static_resource_parameters_s));
+        if (!resource_copy_ptr->static_resource_parameters) {
+            sn_grs_resource_info_free(handle, resource_copy_ptr);
+            return SN_NSDL_FAILURE;
+        }
+
+        memset(resource_copy_ptr->static_resource_parameters, 0, sizeof(sn_nsdl_static_resource_parameters_s));
+        resource_copy_ptr->static_resource_parameters->observable = resource_ptr->static_resource_parameters->observable;
+        resource_copy_ptr->static_resource_parameters->mode = resource_ptr->static_resource_parameters->mode;
+        resource_copy_ptr->static_resource_parameters->external_memory_block = resource_ptr->static_resource_parameters->external_memory_block;
+        resource_copy_ptr->static_resource_parameters->access = resource_ptr->static_resource_parameters->access;
+        resource_copy_ptr->static_resource_parameters->coap_content_type = resource_ptr->static_resource_parameters->coap_content_type;
+
```

Comment:
As a future dev suggestion, temporary variables caching the value of "resource_copy_ptr->static_resource_parameters" and "resource_ptr->static_resource_parameters" might reduce generated code. Or compiler may be able to optimize that itself. At least the source lines might fit to screen;)